geoSegment.js

```diff
@@ -0,0 +1,353 @@
+  let encodedFilename = encodeURIComponent(sessionSeriesId);
+  if (encodedFilename.length > 234) {
+    encodedFilename = encodedFilename.slice(0, 234);
+  }
+  return `./${SESSION_SERIES_DIRECTORY_PATH}/${encodedFilename}.json`;
+}
+
+/**
+ *
+ * @param {RpdeItem<SessionSeriesData>[]} items
+ * @param {Segment[]} segments
+ */
+async function processSessionSeriesItems(items, segments) {
+  items.forEach(async (item) => {
+    // Filter deleted SessionSeries
+    if (item.state === 'deleted') {
+      return;
+    }
+
+    // Filter virtual sessions
+    let sessionSeriesPhysicalLocationGeo;
+    if (item.data.location && item.data.location.geo) {
+      sessionSeriesPhysicalLocationGeo = item.data.location.geo;
+    } else if (item.data['beta:affiliatedLocation'] && item.data['beta:affiliatedLocation'].geo) {
+      sessionSeriesPhysicalLocationGeo = item.data['beta:affiliatedLocation'].geo;
+    } else {
+      return;
+    }
+
+    const sessionSeriesData = {
+      ...item.data,
+      'imin:segment': [],
+    };
+
+    // Add imin:segment if applicable
+    for (const segment of segments) {
+      const segmentRadiusInMeters = segment.radius * 1000;
+      const distanceBetweenGeos = geolib.getDistance(
+        { latitude: sessionSeriesPhysicalLocationGeo.latitude, longitude: sessionSeriesPhysicalLocationGeo.longitude },
+        { latitude: segment.latitude, longitude: segment.longitude },
+      );
+      if (distanceBetweenGeos <= segmentRadiusInMeters) {
+        sessionSeriesData['imin:segment'].push(segment.identifier);
+      }
+    }
+    // If there are no segments, drop the SessionSeries as it will not appear in an output folder and therefore we don't need to store it
+    if (sessionSeriesData['imin:segment'].length === 0) {
+      return;
+    }
+
+    // Validate Schedules and generate ScheduledSessions if needed
+    const scheduledSessionsForEveryEventSchedule = [];
+    const correctedEventSchedules = []; // EventSchedules that are type Schedule but are not valid should be changed to PartialSchedule, and not generate ScheduledSessions
+    if (Array.isArray(sessionSeriesData.eventSchedule)) {
+      for (const eventSchedule of sessionSeriesData.eventSchedule) {
+        if (eventSchedule.type !== SCHEDULE_TYPE) {
+          continue;
+        }
+        // Validate
+        const joiValidation = MINIMAL_EVENT_SCHEDULE_SCHEMA.validate(eventSchedule, {
+          allowUnknown: true,
+        });
+
+        if (joiValidation.error) {
+          const correctedSchedule = {
+            ...eventSchedule,
+            type: 'PartialSchedule',
+          };
+          correctedEventSchedules.push(correctedSchedule);
+          continue;
+        }
+
+        scheduledSessionsForEveryEventSchedule.push(
+          ...generateDatesWithinTimeWindow(eventSchedule),
+        );
+      }
+    }
+
+    // Write SessionSeries to file
+    const sessionSeriesFilePath = getSessionSeriesFilePath(item.id);
```

Comment:
I know that we expect `.id` to always be present in SeS, ScS and SeS-generated-from-Schedule due to our current understanding of Firehose and the idTemplate logic, but I wonder if we should have a quick check just before generating the file paths that just stops processing that item if the `.id` is missing (never know how things might change in the future). Otherwise, may produce bizarre files (`undefined.json` or just `.json`?)

Up to you. Maybe I'm being too paranoid 😅

Reply:
Have null checked